src/ray/core_worker/core_worker.cc

```diff
@@ -1412,7 +1421,10 @@ void CoreWorker::SubmitActorTask(const ActorID &actor_id, const RayFunction &fun
       worker_context_.GetCurrentJobID(), worker_context_.GetCurrentTaskID(),
       next_task_index, actor_handle->GetActorID());
   const std::unordered_map<std::string, double> required_resources;
-  BuildCommonTaskSpec(builder, actor_handle->CreationJobID(), actor_task_id,
+  auto task_name = task_options.name.empty()
+                       ? function.GetFunctionDescriptor()->DefaultTaskName()
```

Comment:
It's probably a small overhead, but I wonder if we should just leave the task name as empty and try to derive it as before, instead of computing it and saving it as part of the descriptor.

Perhaps try `ray microbenchmark` on a large machine and see if there's any measurable degradation in performance?

Reply:
I thought about going that route, but I felt like it made the task spec abstraction messier. If we know the task name at submission time (either user-provided or derived from the function descriptor), why not set it on the task spec? If we set it, all downstream consumers of the task spec will know that `task_spec.GetName()` will give them a non-empty user-friendly task name, which is a clean contract. If we don't set it, all downstream consumers will have to know that the task name field might be empty, in which case the consumer is expected to construct a user-friendly task name off of the function descriptor, which seems less clean and will duplicate that logic at every point in which the task name is consumed.

If something downstream wanted to be able to delineate between a user-provided task name and a task name derived from the function descriptor (e.g. so they can construct a task name that makes the most sense for their context that's different than the default one generated by the function descriptor), _that_ I could see as an argument to leave the task name empty if not given by the user. Can you see any cases like that popping up?

Reply:
Another option, if the uncommon case arises in which the default task name isn't what a consumer wants when the user doesn't specify a task name, would be to move this logic to `TaskSpec::GetName()`: if `!name_.empty()`, return `name_`, otherwise return `GetFunctionDescriptor()->DefaultTaskName()`. That would make `SubmitTask` and `SubmitActorTask` cleaner, and we could expose something like `TaskSpec::IsNameUserProvided() { return !name_.empty(); }` so consumers can delineate between user-specified and default task names. Each language that wraps the `TaskSpecification` proto message would have to implement this logic, but given that this is an internal proto message, that seems reasonable.

I'm hoping to preserve the happy common path, where most task spec consumers are don't care whether the task name was user-provided or the default derived from the function descriptor: they're just happy to have a human-readable task name.